### PR TITLE
Ensure elytron is on the classpath before using it

### PR DESCRIPTION
**Pull request by lampajr**
**Thank you for submitting this pull request**

**JIRA**: 

- https://issues.redhat.com/browse/RHPAM-4714

**referenced Pull Requests**: 
* https://github.com/kiegroup/droolsjbpm-integration/pull/2961

Changes performed https://github.com/kiegroup/droolsjbpm-integration/pull/2961 were not enough as kieServerMatrix tests are still failing while using `tomcat` server:

```
[INFO] [talledLocalContainer] SEVERE: Exception sending context initialized event to listener instance of class [org.kie.server.remote.rest.common.Bootstrap]
[INFO] [talledLocalContainer] java.lang.NoClassDefFoundError: org/wildfly/security/auth/server/SecurityDomain
[INFO] [talledLocalContainer] 	at org.kie.server.services.impl.security.ElytronIdentityProvider.available(ElytronIdentityProvider.java:33)
[INFO] [talledLocalContainer] 	at org.kie.server.services.impl.KieServerImpl.init(KieServerImpl.java:157)
[INFO] [talledLocalContainer] 	at org.kie.server.services.impl.KieServerImpl.init(KieServerImpl.java:142)
[INFO] [talledLocalContainer] 	at org.kie.server.services.impl.KieServerLocator$KieServerLocatorHelper.<clinit>(KieServerLocator.java:25)
[INFO] [talledLocalContainer] 	at org.kie.server.services.impl.KieServerLocator.getInstance(KieServerLocator.java:32)
[INFO] [talledLocalContainer] 	at org.kie.server.remote.rest.common.Bootstrap.contextInitialized(Bootstrap.java:51)
[INFO] [talledLocalContainer] 	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4682)
[INFO] [talledLocalContainer] 	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5143)
....
[INFO] [talledLocalContainer] Caused by: java.lang.ClassNotFoundException: org.wildfly.security.auth.server.SecurityDomain
[INFO] [talledLocalContainer] 	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1365)
[INFO] [talledLocalContainer] 	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1188)
[INFO] [talledLocalContainer] 	... 47 more
```

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

**Comment by lampajr** [comment]
jenkins do fdb

**Comment by lampajr** [comment]
Jenkins run fdb

**Comment by lampajr** [comment]
jenkins retest this

File: kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/security/adapters/JMSSecurityAdapter.java [hunk 1]
```diff
@@ -105,14 +113,22 @@ private static UserDetails getUserDetails(final String user, final String pass)
...
         final List<String> roles = new ArrayList<>();
 
-        final SecurityIdentity identity = SecurityDomain.getCurrent().authenticate(user, new PasswordGuessEvidence(password.toCharArray()));
-        userDetails.setName(identity.getPrincipal().getName());
-        for (String role : identity.getRoles()) {
-            roles.add(role);
+        if (isElytronActive) {
+            try {
+                final SecurityIdentity identity = SecurityDomain.getCurrent().authenticate(user,
+                        (Evidence) passwordGuessEvidenceClass.getDeclaredConstructor(char[].class).newInstance(password.toCharArray()));
```
Comment (lampajr) [inline kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/security/adapters/JMSSecurityAdapter.java:123]: Without this workaround, `kie-server-tests` were still failing complaining about `Evidence` class:
```
[INFO] [talledLocalContainer] java.util.ServiceConfigurationError: org.kie.server.api.security.SecurityAdapter: org.kie.server.services.impl.security.adapters.JMSSecurityAdapter Unable to get public no-arg constructor
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:582)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:673)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1233)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
...
 [INFO] [talledLocalContainer] Caused by: java.lang.NoClassDefFoundError: org/wildfly/security/evidence/Evidence
 [INFO] [talledLocalContainer]   at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
 [INFO] [talledLocalContainer]   at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:3137)
 [INFO] [talledLocalContainer]   at java.base/java.lang.Class.getConstructor0(Class.java:3342)
 [INFO] [talledLocalContainer]   at java.base/java.lang.Class.getConstructor(Class.java:2151)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:660)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:657)
 [INFO] [talledLocalContainer]   at java.base/java.security.AccessController.doPrivileged(Native Method)
 [INFO] [talledLocalContainer]   at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:668)
 [INFO] [talledLocalContainer]   ... 52 more
 [INFO] [talledLocalContainer] Caused by: java.lang.ClassNotFoundException: org.wildfly.security.evidence.Evidence
 [INFO] [talledLocalContainer]   at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1365)
 [INFO] [talledLocalContainer]   at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1188)
 [INFO] [talledLocalContainer]   ... 60 more
```

Tests were ran using following command:
```
mvn -f kie-server-parent/kie-server-tests/pom.xml -Dcontainer=tomcat9 -Dcontainer.startstop.timeout.millis=240000 -Deap7.download.url=http://download.devel.redhat.com/released/JBoss-middleware/eap7/7.4.10/jboss-eap-7.4.10.zip -Dmaven.test.failure.ignore=true -Ddeployment.timeout.millis=240000 -B -e -fae -nsu verify -Ptomcat9
```

**Comment by lampajr** [comment]
`Linux - Pull Request` there are two test failures (actually one error and one failure):
```
[2023-05-30T12:36:10.550Z] [ERROR] Failures: 
[2023-05-30T12:36:10.550Z] [ERROR]   ContainerJmsResponseHandlerIntegrationTest.testContainerWithAsyncResponseHandler:110 
[2023-05-30T12:36:10.550Z] Expected size:<1> but was:<2> in:
[2023-05-30T12:36:10.550Z] <[KieContainerResource [containerId=asyncContainer, releaseId=org.kie.server.testing:container-crud-tests1:2.1.0.GA, resolvedReleaseId=null, status=STARTED],
[2023-05-30T12:36:10.550Z]     KieContainerResource [containerId=asyncScannerContainer, releaseId=org.kie.server.testing:container-crud-tests1:2.1.0.GA, resolvedReleaseId=null, status=STARTED]]>
[2023-05-30T12:36:10.550Z] [ERROR] Errors: 
[2023-05-30T12:36:10.550Z] [ERROR]   ContainerJmsResponseHandlerIntegrationTest.testScannerWithAsyncResponseHandler:154->KieServerBaseIntegrationTest.createContainer:137->KieServerBaseIntegrationTest.createContainer:141->KieServerBaseIntegrationTest.createContainer:150 » NoEndpointFound
[2023-05-30T12:36:10.550Z] [INFO] 
[2023-05-30T12:36:10.550Z] [ERROR] Tests run: 130, Failures: 1, Errors: 1, Skipped: 0
```

they could be related to the change, re-running to avoid temp issues..

jenkins retest this

**Comment by lampajr** [comment]
jenkins do fdb

**Comment by lampajr** [comment]
jenkins retest this

**Comment by lampajr** [comment]
jenkins retest this

File: kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/security/adapters/JMSSecurityAdapter.java (hunk 1 above)
Comment (lampajr) [reply kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/security/adapters/JMSSecurityAdapter.java:123]: Mmm ok it seems there are still some other parts that I missed, checking..

**Comment by lampajr** [comment]
Jenkins run fdb

**Comment by sutaakar** [comment]
@lampajr Just one minor point, can you fix the bug reported by Sonarcloud?
Otherwise looks ok (didn't try to run it by myself).

**Comment by lampajr** [comment]
@sutaakar yeah I am trying to refactor it in a better way, I think I found how to do it

**Comment by lampajr** [comment]
> @lampajr Just one minor point, can you fix the bug reported by Sonarcloud? Otherwise looks ok (didn't try to run it by myself).

The bug is fixed now, still low level of coverage :/

**Comment by mareknovotny** [comment]
it seems we need to wait until we can use our installation host for node

**Review by sutaakar — APPROVED** [review]
Looks ok, though it needs to be checked in nightly jobs for all application servers after merge.

**Comment by lampajr** [comment]
@sutaakar @mareknovotny added some new unit tests trying to cover some branches https://github.com/kiegroup/droolsjbpm-integration/pull/2967/commits/f437fda81c2284779c52607da0fc0c2c17885f61, wdyt?

**Comment by sutaakar** [comment]
Looks ok for me

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/test/java/org/kie/server/services/jbpm/security/ElytronUserGroupAdapterTest.java [hunk 2]
```diff
@@ -95,4 +106,43 @@ public void testSecurityForLoggedUser() {
...
+        when(adapter.getUserName()).thenReturn(WRONG_USER_ID);
+        when(adapter.runAsPrincipalExists(USER_ID)).thenThrow(new RealmUnavailableException());
+        lenient().when(adapter.toPrincipalRoles(Mockito.anyString())).thenReturn(Arrays.asList(ROLE_1, ROLE_2, ROLE_3));
+        lenient().when(adapter.toRunAsPrincipalRoles(Mockito.anyString(), Mockito.eq(false))).thenReturn(Arrays.asList(ROLE_1, ROLE_2, ROLE_3));
+
+        List<String> roles = adapter.getGroupsForUser(USER_ID);
+
+        Assertions.assertThat(roles)
+                .isNotNull()
+                .hasSize(0);
```
Comment (mareknovotny) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/test/java/org/kie/server/services/jbpm/security/ElytronUserGroupAdapterTest.java:124]:
```suggestion
                .isEmpty();
```
from Sonar cloud suggestions to have it simplified ;)

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/security/ElytronUserGroupAdapter.java [hunk 3]
```diff
@@ -23,44 +23,63 @@
...
+            logger.info("Unable to find org.wildfly.security.authz.AuthorizationFailureException, disabling elytron adapter");
+        }
+    }
+
+    protected boolean isActive() {
+        return this.authorizationFailureExceptionClass != null;
+    }
 
     @Override
     public List<String> getGroupsForUser(String userId) {
```
Comment (mareknovotny) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/security/ElytronUserGroupAdapter.java:52]: `Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed.` https://sonarcloud.io/project/issues?resolved=false&severities=CRITICAL&pullRequest=2967&id=org.drools%3Adroolsjbpm-integration&open=AYhzEjM4inEykiRLLCBi

**Comment by lampajr** [comment]
Coverage improved a bit (from 60% to 70%) but still complaining about `Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed.` I'll try to refactor again that method in order to reduce the complexity again

**Comment by mareknovotny** [comment]
i think you can run sonar with sonarlint also locally in your IDE if you need to check what improvements value it is ;)

**Review by mareknovotny — APPROVED** [review]

**Pull request merged**